File: atsa/src/main/java/jp/kaiz/atsassistmod/controller/trainprotection/RnATSController.java
package jp.kaiz.atsassistmod.controller.trainprotection;

import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity;

/**
 * Rn-ATS uses the signal aspect ahead of the train through RTM's legacy
 * {@code getSignal()} compatibility API.
 */
public class RnATSController extends TrainProtection {
    private int limitSpeed;

    @Override
    public void onTick(TrainEntity train, double distance) throws Exception {
        super.onTick(train, distance);
        switch (train.getSignal()) {
            case 1 -> this.limitSpeed = 0;
            case 2 -> this.limitSpeed = 15;
            case 3 -> this.limitSpeed = 25;
            case 4 -> this.limitSpeed = 35;
            case 5 -> this.limitSpeed = 45;
            case 6 -> this.limitSpeed = 55;
            case 7 -> this.limitSpeed = 65;
            case 8 -> this.limitSpeed = 75;
            case 9 -> this.limitSpeed = 85;
            case 10 -> this.limitSpeed = 95;
            case 11 -> this.limitSpeed = 100;
            case 12 -> this.limitSpeed = 110;
            case 13 -> this.limitSpeed = 120;
            case 14 -> this.limitSpeed = 130;
            default -> this.limitSpeed = 25;
        }
    }

    @Override
    public int getNotch(float speedH) {
        float overSpeed = speedH - this.limitSpeed;
        if (overSpeed > 5) {
            return -7;
        } else if (overSpeed > 0 || this.train.getSignal() == 1) {
            return -4;
        } else {
            return 1;
        }
    }

    @Override
    public TrainProtectionType getType() {
        return TrainProtectionType.RnATS;
    }

    @Override
    public int getDisplaySpeed() {
        return this.limitSpeed;
    }
}

File: atsa/src/main/kotlin/jp/kaiz/atsassistmod/controller/trainprotection/RnATSController.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package jp.kaiz.atsassistmod.controller.trainprotection

import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity

/** Rn-ATS uses the signal aspect ahead of the train through RTM's legacy getSignal compatibility API. */
open class RnATSController : TrainProtection() {
    private var limitSpeed = 0

    @Throws(Exception::class)
    override fun onTick(train: TrainEntity, distance: Double) {
        super.onTick(train, distance)
        limitSpeed = when (train.signal) {
            1 -> 0
            2 -> 15
            3 -> 25
            4 -> 35
            5 -> 45
            6 -> 55
            7 -> 65
            8 -> 75
            9 -> 85
            10 -> 95
            11 -> 100
            12 -> 110
            13 -> 120
            14 -> 130
            else -> 25
        }
    }

    override fun getNotch(speedH: Float): Int {
        val overSpeed = speedH - limitSpeed
        return when {
            overSpeed > 5 -> -7
            overSpeed > 0 || train?.signal == 1 -> -4
            else -> 1
        }
    }

    override fun getType(): TrainProtectionType = TrainProtectionType.RnATS

    override fun getDisplaySpeed(): Int = limitSpeed
}
